File: python/arrays-and-hashing/top-k-frequent-elements.py
import heapq


class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # Create finalanswer array
        finalanswer = []
        # Hash Map to keep count of the number of elements equal to the key
        # Key: Numbers, Value: Count of Numbe
        count = defaultdict(int)

        # O(n) loop to keep track of the count of each number
        for i in nums:
            count[i] += 1

        # Heap to arrange K most frequent
        heap = []

        # O(n) loop to populate the heap based on count, with the highest count being at the top of the heap
        for i in count.keys():
            heapq.heappush(heap, (-count[i], i))

        # Popping the K most frequent elements from the heap and appending them to the final array
        for i in range(k):
            finalanswer.append(heapq.heappop(heap)[1])

        return finalanswer

File: c++/arrays-and-hashing/top-k-frequent-elements.cc
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        std::priority_queue<std::tuple<int, int>> pq;
        std::unordered_map<int, int> count;

        for (int num : nums) {
            count[num] += 1;
        }

        for (auto [num, counter] : count) {
            pq.emplace(counter, num);
        }

        std::vector<int> answer;
        for (int i = 0; i < k; ++i) {
            answer.push_back(std::get<1>(pq.top()));
            pq.pop();
        }
        return answer;
    }
};
